stonesoup/updater/kalman.py

```diff
@@ -562,3 +558,88 @@ def _posterior_covariance(self, hypothesis):
                  hypothesis.measurement_prediction.cross_covar.T)
 
         return post_cov, kalman_gain
+
+
+class IteratedKalmanUpdater(ExtendedKalmanUpdater):
+    r"""This version of the Kalman updater runs an iteration over the linearisation of the
+    sensor function in order to refine the posterior state estimate. Specifically,
+
+    .. math::
+
+        x_{k,i+1} &= x_{k|k-1} + K_i(z - h(x_{k,i}) - H_i (x_{k|k-1} - x_{k,i}))
+
+        P_{k,i+1} &= (I - K_i H_i) P_{k|k-1}
+
+    where,
+
+    .. math::
+
+        H_i &= h^{\prime}(x_{k,i}), \ K_i = P_{k|k-1} H_i^T (H_i P_{k|k-1} H_i^T + R)^{-1}
+
+    and
+
+    .. math::
+
+        x_{k,0} &= x_{k|k-1} \ \mathrm{and} \ P_{k,0} = P_{k|k-1}
+
+    It inherits from the ExtendedKalmanUpdater as it uses the same linearisation of the sensor
+    function via the :meth:`_measurement_matrix()` function.
+    """
+
+    tolerance = Property(float, default=1e-6,
+                         doc="The value of the difference in the measure used as a stopping "
+                             "criterion.")
+    measure = Property(Measure, default=Euclidean(), doc="")
+
+    def update(self, hypothesis, **kwargs):
+        r"""The iterated Kalman update method. Given a hypothesised association between a predicted
+        state or predicted measurement and an actual measurement,
+        calculate the posterior state.
+
+        Parameters
+        ----------
+        hypothesis : :class:`~.SingleHypothesis`
+            the prediction-measurement association hypothesis. This hypothesis
+            may carry a predicted measurement, or a predicted state. In the
+            latter case a predicted measurement will be calculated.
+        **kwargs : various
+            These are passed to the measurement model function
+
+        Returns
+        -------
+        : :class:`~.GaussianStateUpdate`
+            The posterior state Gaussian with mean :math:`\mathbf{x}_{k|k}` and
+            covariance :math:`P_{x|x}`
+
+        """
+
+        # Record the starting point
+        prev_state = hypothesis.prediction
+
+        # Get the measurement model
+        measurement_model = self._check_measurement_model(hypothesis.measurement.measurement_model)
+
+        # The first iteration is just the application of the EKF
+        post_state = super().update(hypothesis, **kwargs)
+
+        # Now update the measurement prediction mean and loop
+        while self.measure(prev_state, post_state) > self.tolerance:
```

Comment:
I think it is worth having a max iteration count just in case.  Something like this should work:
```python
iterations = 0
while self.measure(prev_state, post_state) > self.tolerance \
        and iterations < self.max_iterations:
    # Do stuff here...
    iterations += 1
```

or maybe:
```python
iterations = 0
while self.measure(prev_state, post_state) > self.tolerance:
    # Do stuff here...
    iterations += 1
    if iterations > self.max_iterations:
        warnings.warn("Iterated Kalman update did not converge")
        break
```

Reply:
That simple eh? Would have expected something a bit fancier from Python.

Reply:
Done

Reply:
The second way